### PR TITLE
form with image dragging + dashboard improvingform with image dragging + dashboard improving

### PR DESCRIPTION
**Pull request by devisable**

### CODE REVIEW

**Review by devisable — COMMENTED** [review]
form with image dragging + dashboard improvingform with image dragging + dashboard improving

**Comment by jsru** [comment]
Добавляю преподавателя (@dosandk) для код-ревью.

File: 09-tests-for-frontend-apps/1-product-form-v2/index.js [hunk 1]
```diff
@@ -1,16 +1,285 @@
...
+    this.url = new URL(`/api/rest/products`, BACKEND_URL);
+    this.data = {};    
   }
 
   async render () {
 
+    const categoriesPr = this.loadSubcategories();
+    const productPr = this.productId ? this.loadData(this.productId) : Promise.resolve([this.defaultFormData]);
+
+    const [subcategories, productResponse] = await Promise.all([categoriesPr, productPr]);
```
Comment (dosandk) [inline 09-tests-for-frontend-apps/1-product-form-v2/index.js:36]: 👍 good

File: 09-tests-for-frontend-apps/1-product-form-v2/index.js [hunk 2]
```diff
@@ -1,16 +1,285 @@
...
+
+    inputFile.click();
+  };
+
+  async loadData(id) {
+    
+    this.url.searchParams.set("id", id);    
+    
+    try {
+      return await fetchJson(this.url);
```
Comment (dosandk) [inline 09-tests-for-frontend-apps/1-product-form-v2/index.js:108]: 👍 good

File: 09-tests-for-frontend-apps/1-product-form-v2/index.js [hunk 3]
```diff
@@ -1,16 +1,285 @@
...
+      </form>
+    </div>`;
+  }
+  
+  getSubcategories(){
+
+    let i = 0;
+    this.subcategories.map(category => {
+      category.subcategories.map(item => {
+        this.subElements.subcategory.options[i++] = new Option(`${category.title} > ${item.title}`, item.id);
```
Comment (dosandk) [inline 09-tests-for-frontend-apps/1-product-form-v2/index.js:184]: у метода `map` есть второй аргумент - `index`

**Review by dosandk — APPROVED** [review]
good 👍

**Pull request merged**